iothub_client/src/iothub_client_ll_uploadtoblob.c

```diff
@@ -320,584 +481,730 @@ static int IoTHubClient_LL_UploadToBlob_step1and2(IOTHUB_CLIENT_LL_UPLOADTOBLOB_
-                                    LogError("unable to HTTPAPIEX_ExecuteRequest");
                                 }
                                 else
                                 {
-                                    wasIoTHubRequestSuccess = 1;
+                                    if (parseResultFromIoTHub(STRING_c_str(responseAsString), correlationId, sasUri) != 0)
+                                    {
+                                        LogError("unable to parse json result");
+                                        result = MU_FAILURE;
+                                    }
+                                    else
+                                    {
+                                        result = 0;
+                                    }
+
+                                    STRING_delete(responseAsString);
                                 }
-                                break;
-                            }
-                        }
-                        if (wasIoTHubRequestSuccess == 1)
-                        {
-                            const unsigned char*responseContent_u_char = BUFFER_u_char(responseContent);
-                            size_t responseContent_length = BUFFER_length(responseContent);
-                            STRING_HANDLE responseAsString = STRING_from_byte_array(responseContent_u_char, responseContent_length);
-                            if (responseAsString == NULL)
-                            {
-                                result = MU_FAILURE;
-                                LogError("unable to get the response as string");
                             }
                             else
                             {
-                                if (parse_result_json(STRING_c_str(responseAsString), correlationId, sasUri) != 0)
-                                {
-                                    LogError("unable to parse json result");
-                                    result = MU_FAILURE;
-                                }
-                                else
-                                {
-                                    result = 0;
-                                }
-                                STRING_delete(responseAsString);
+                                result = MU_FAILURE;
                             }
+
+                            HTTPHeaders_Free(iotHubRequestHttpHeaders);
                         }
+
+                        HTTPAPIEX_Destroy(iotHubHttpApiExHandle);
                     }
+
                     BUFFER_delete(responseContent);
                 }
+
                 BUFFER_delete(blobBuffer);
             }
+
             STRING_delete(blobName);
         }
+
         STRING_delete(relativePath);
     }
+
+    return result;
+}
+
+static int IoTHubClient_LL_UploadToBlob_NotifyIoTHubOfUploadCompletion(IOTHUB_CLIENT_LL_UPLOADTOBLOB_HANDLE_DATA* upload_data, BUFFER_HANDLE messageBody)
```

Comment:
How does this handle the situation where the upload has already timed out and the hub doesn't have the reference to the correlation id anymore? (Does that get handled the same way as a success?)

Reply:
The function only returns success if the Rest API call to IoT Hub succeeds (http status code >= 200 and < 300). If the upload has already timed out in the hub, it will fail the http request and the function will also fail.

Reply:
In that case, since this doesn't need to keep being retried, should it be considered a success/a different scenario so it doesn't seem like it should get retried?

Reply:
Alright, if trying to send a completion notification for a correlation id that is no longer valid results in this response:

```log
Error: Time:Wed Jul 19 21:13:03 2023 File:/home/ewertons/code/s2/azure-iot-sdk-c/c-utility/adapters/httpapi_curl.c Func:HTTPAPI_ExecuteRequest Line:778 Failure in HTTP communication: server reply code is 400
Info: HTTP Response:{"Message":"{\"errorCode\":400015,\"trackingId\":\"ed0801cda12f4ae2b827ea574a726e3e-G:0-TimeStamp:07/20/2023 04:13:03\",\"message\":\"Invalid correlation ID\",\"timestampUtc\":\"2023-07-20T04:13:03.3175015Z\"}","ExceptionMessage":""}
Error: Time:Wed Jul 19 21:13:03 2023 File:/home/ewertons/code/s2/azure-iot-sdk-c/iothub_client/src/iothub_client_ll_uploadtoblob.c Func:send_http_sas_request Line:131 HTTP failed response code was 400
```

I tried looking for documentation about IoT Hub upload to blob API error codes, and without surprise there is nothing out there. Customers even reached out looking for that info, but without much luck.
https://learn.microsoft.com/en-us/answers/questions/483211/list-of-error-codes-for-iothub-file-upload-api

So in this case we do not have a way to distinguish if the hub failed because there is a malformed request (HTTP 400) or if the upload correlation id could not be found (also, HTTP 400).
The correct thing would be for the hub to return 404 when the upload is no longer in its list.
I'm afraid we do not have a way to improve the return of `IoTHubClient_LL_UploadToBlob_NotifyIoTHubOfUploadCompletion`

Reply:
A few questions:
- Do we get a 400 if the request completely fails too (ex. no network), or something different?
- Is a malformed request (400) recoverable by resending, or is this a non-retriable state as well?
In my mind, I see this flow working (obviously simplified in the actual code):
if (http success)
    remove correlation id from pending list
if (http failure that's retriable (ex. no network)
    keep correlation id in pending list and try again later
if (correlation id can't be found (400))
   remove correlation id from pending list since it's not taking a place in the hub's throttling limit anymore and probably timed out - essentially a success case
if (malformed request)
   would we also remove the correlation id from the pending list here because whatever we sent doesn't get parsed properly by the hub anyways, so sending the message again will give the same response? Or is this likely a case of message corruption and we should try again?

Reply:
As we spoke in person, this is mostly handled by the retry in the iothub_client_sample_upload_to_blob_custom.
Unfortunately IoT Hub error codes for u2b are not fully documented and it would be not actionable for customers to get those codes from the SDK api anyway.